kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/contexts/AsyncContextRouter.java

```diff
@@ -0,0 +1,273 @@
+ *  which should always have a handle on the specific context instance and pass things
+ *  off directly between if needed. In other words, this class is simply made to act
+ *  as an interface between the user and the underlying context objects. It is only
+ *  passed in to the user's #init and only serves to redirect the ProcessorContext
+ *  methods that the user invokes from their processor (whether in #init, #process,
+ *  or #close)
+ */
+@SuppressWarnings("checkstyle:overloadmethodsdeclarationorder")
+public class AsyncContextRouter<KOut, VOut>
+    implements ProcessorContext<KOut, VOut>, FixedKeyProcessorContext<KOut, VOut> {
+
+  private final Logger log;
+
+  private final int partition;
+
+  // Flipped to true once regular processing begins, ie between when #init and #close
+  // are invoked on the AsyncProcessor.
+  // When true, the StreamThread should be executing and its context used to delegate
+  // When false, only AsyncThreads should be accessing this router
+  private final AtomicBoolean isInProcessingMode = new AtomicBoolean(false);
+
+  private final StreamThreadProcessorContext<KOut, VOut> streamThreadProcessorContext;
+
+  public AsyncContextRouter(
+      final String logPrefix,
+      final int partition,
+      final StreamThreadProcessorContext<KOut, VOut> streamThreadProcessorContext
+  ) {
+    this.log = new LogContext(logPrefix).logger(AsyncContextRouter.class);
+    this.partition = partition;
+
+    this.streamThreadProcessorContext = streamThreadProcessorContext;
+  }
+
+  public void startProcessingMode() {
+    isInProcessingMode.setOpaque(true);
+  }
+
+  public void endProcessingMode() {
+    isInProcessingMode.setOpaque(false);
+  }
+
+  /**
+   * Look up the appropriate context based on whether processing mode is on
+   */
+  private ProcessorContext<KOut, VOut> lookupContext() {
+    if (isInProcessingMode.getOpaque()) {
+      return lookupContextForAsyncThread();
+    } else {
+      return streamThreadProcessorContext;
+    }
+  }
+
+  /**
+   * Look up the appropriate context based on whether processing mode is on,
+   * and return it as a fixed-key context. This is needed because unfortunately
+   * there is no ProcessorContext interface with all #forward overloads beneath
+   * and so we have to split this method up into two versions, even though they
+   * do the exact same thing inside.
+   */
+  private FixedKeyProcessorContext<KOut, VOut> lookupFixedKeyContext() {
+    if (isInProcessingMode.getOpaque()) {
+      return lookupContextForAsyncThread();
+    } else {
+      return streamThreadProcessorContext;
+    }
+  }
+
+  /**
+   * Look up the context when you know it should be an AsyncThreadProcessorContext.
+   * This method includes a safety check to verify that assumption
+   */
+  private AsyncThreadProcessorContext<KOut, VOut> lookupContextForAsyncThread() {
+    if (Thread.currentThread() instanceof AsyncThread) {
```

Comment:
we should make sure to preserve this assertion if we go the ThreadLocal route

Reply:
Take a look at the new AsyncUserProcessorContext class (slight renaming of the AsyncProcessorContext that you wrote), I added some safety checks but lmk if you feel they're sufficient?

They certainly helped me debug the refactor lol